internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -267,67 +265,207 @@ func writeKubeadmConfig(data *userdata.UserData) (err error) {
+// security policies necessary for kubeadm init to function.
+// This serves as a base for the list of files that need to
+// be synced via trustd from other nodes
+func requiredFiles() []string {
+	return []string{
+		constants.AuditPolicyPathInitramfs,
+		constants.EncryptionConfigInitramfsPath,
+		constants.KubeadmCACert,
+		constants.KubeadmCAKey,
+		constants.KubeadmSACert,
+		constants.KubeadmSAKey,
+		constants.KubeadmFrontProxyCACert,
+		constants.KubeadmFrontProxyCAKey,
+		constants.KubeadmEtcdCACert,
+		constants.KubeadmEtcdCAKey,
+	}
+}
+
+// fileSet compares the list of required files to the ones
+// already present on the node and returns the delta
+func fileSet() []*proto.ReadFileRequest {
+	files := requiredFiles()
+
+	fileRequests := []*proto.ReadFileRequest{}
+	// Check to see if we already have the file locally
+	for _, file := range files {
+		if _, err := os.Stat(file); os.IsNotExist(err) {
+			fileRequests = append(fileRequests, &proto.ReadFileRequest{Path: file})
 		}
-		if err = ioutil.WriteFile(key, data.Key, 0400); err != nil {
-			return fmt.Errorf("write %s: %v", key, err)
+	}
+
+	return fileRequests
+}
+
+// createTrustdClients handles instantiating a trustd client connection
+// to each trustd endpoint defined in userdata
+func createTrustdClients(data *userdata.UserData) ([]proto.TrustdClient, error) {
+	var creds basic.Credentials
+	var err error
+
+	trustds := []proto.TrustdClient{}
+
+	creds, err = basic.NewCredentials(data.Services.Trustd)
+	if err != nil {
+		return trustds, err
+	}
+
+	// Create a trustd client for each endpoint to set up
+	// a fan out approach to gathering the files
+	var conn *grpc.ClientConn
+	for _, endpoint := range data.Services.Trustd.Endpoints {
+		conn, err = basic.NewConnection(endpoint, constants.TrustdPort, creds)
+		if err != nil {
+			return trustds, err
 		}
+		trustds = append(trustds, proto.NewTrustdClient(conn))
 	}
-	return nil
+	return trustds, nil
 }
 
-func writeFiles(client proto.TrustdClient, files []string) (err error) {
-	errChan := make(chan error)
-	doneChan := make(chan bool)
-	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Minute)
-	defer cancelFunc()
-
-	go func() {
-		<-ctx.Done()
-		errChan <- ctx.Err()
-	}()
-
-	go func() {
-		var err error
-		for _, f := range files {
-		L:
-			req := &proto.ReadFileRequest{
-				Path: f,
+func download(client proto.TrustdClient, file *proto.ReadFileRequest, content chan<- []byte, donezo <-chan struct{}) {
+	var resp *proto.ReadFileResponse
+	var err error
+	attempt := 0
+	maxWait := float64(64)
+
+	// Wait indefinitely with a backoff
+	for {
+		attempt++
+
+		select {
+		case <-donezo:
+			return
+		default:
+		}
+
+		if resp, err = client.ReadFile(context.Background(), file); err == nil {
+			select {
+			case <-donezo:
+				return
+			case content <- resp.Data:
+			case <-time.After(2 * time.Second):
+				// retry
+				// help with expiration of unnecessary requests
```

Comment:
I think this pattern could be simplified with the use of contexts:

`donezo` looks like generic "abort" channel which can be replaced with cancellable contxt
`time.After` sounds like context abort

closer to code:

`download(ctx context.Context, ...)`

inside the function you can create sub-context with timeout:

```
subctx, subctxCancel := context.WithTimeout(ctx, 2 * time.Second)
defer subctxCancel() // this should probaly go into nested function

resp, err := client.ReadFile(subctx, file)
```

this would abort grpc request either when parent context is cancelled or when timeout passes

Reply:
so I think probably logic might be:

1. `download(ctx)` <- receives context upstream, which can be used to abort whole thing (e.g. on node shutdown)
2. `download(ctx)` schedules N goroutines per endpoint to download given file, they all write results to the same channel, we can have sub-context with timeout to limit time per download attempt
3. if any of the goroutines returns result, abort all others by cancelling sub-ctx
4. if all the goroutines fail to deliver result, go to (2) to retry
5. in the end function returns either failure (context canceled) or file downloaded
6. write function writes down the file.

(so flow is sequential from download to write, but download internally is parallel)

Reply:
These are good points. I have historically always done something like the pattern that Brad implemented, but  a lot of what is implemented can be done with `context`. That will clean up the code quite a bit.